### PR TITLE
Enhancing handler function

### PR DESCRIPTION
**Pull request by MuhammadMouradG**
## Describe your changes in detail
This PR will enhancing the behavior of `_handler` and `wrapped_handler` functions as the discussion in #1098 concludes.

Also, I fixed the test cases to adapt with our edits.

## PR Checklist:
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: src/core/tests/widgets/test_button.py [hunk 1]
```diff
@@ -38,8 +38,8 @@ def callback(widget, **extra):
         self.btn.on_press = callback
         self.assertEqual(self.btn.on_press._raw, callback)
         self.assertEqual(
-            self.btn.on_press('widget', a=1),
-            "called <class 'toga.widgets.button.Button'> with {'a': 1}"
+            self.btn.on_press(self, a=1),
+            "called <class 'tests.widgets.test_button.ButtonTests'> with {'a': 1}"
```
Comment (freakboy3742) [inline src/core/tests/widgets/test_button.py:42]: This is a red flag for me. It's not necessarily *wrong*, but I want to understand why the change exists. 

The external behavior of handlers shouldn't be affected by this change. It should be an entirely internal change to the way the wrapper is bound. So why are the tests changing in this way?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Thanks for the PR; it looks exactly like what I described in the other ticket - but the changes to the tests concern me; details inline. 

To be clear - it could be entirely my fault - It's possible I missed something in my analysis when I described how the handlers could be simplified.

File: src/core/tests/widgets/test_button.py (hunk 1 above)
Comment (MuhammadMouradG) [reply src/core/tests/widgets/test_button.py]: I committed a new fix commit that pass to the `handler` a correct used toga widget. 

Now, and as I understood,
```python
    #  src/core/toga/handlers.py
    def _handler(widget, *args, **kwargs):
         ...
-        result = handler(interface, *args, **kwargs)
+        result = handler(widget, *args, **kwargs)
         ...
```
In the previous implementation of `wrapped_handler`, when we call
```python
self.btn.on_press('widget', a=1)
```
the function `_handler` is called, and the formal parameters in it are replaced with the actual parameters as the following:
`widget='widget'`, `args=*args` and `kwargs={a: 1}`, after that the `_handler` function returns the parameter `result`. Actually, this parameter is equal to the calling of `handler` function with mapping of formal parameters `args` and `kwargs` only with passed parameter from `_handler` and as you note `widget` parameter of `_handler` does not passed to `handler` function.

Unlike the previous implementation in the current implementation the `widget` argument is important and passed to `handler` function and it is the widget that defines this handler and it is the `interface` argument in the previous implementation.

This is what I understood from code, it seems our discussion missed something in analyzing the code.
What do you think?

Comment (MuhammadMouradG) [reply src/core/tests/widgets/test_button.py]: @freakboy3742 What do you think here?

Comment (freakboy3742) [reply src/core/tests/widgets/test_button.py]: I think you're analysis of the old test cases is correct. The old tests (the ones where we are passing in `'widget'`) were effectively ignoring that argument, and getting the "bound" widget when the handler was invoked. By removing the duplicated argument, we've 

However, seeing code like:
```
self.btn.on_press(self.btn, a=1)
```
makes me think that we've missed something along the way. When you invoke `self.method()`, the first (implied) argument to `method()` is `self`. Invoking `self.method(self)` is redundant, unless there's a use case where the first argument *won't* be self. In this case, we're invoking a handler (like `on_press`) because button has been pressed; 

I'm starting to wonder if we've possibly removed the wrong argument - that the argument to `wrapped_handler` is the one we need to preserve, and the argument to `_handler` (and the usage when the handler is actually invoked) are the usages that are redundant.

Does that make sense?

Comment (MuhammadMouradG) [reply src/core/tests/widgets/test_button.py]: It makes very sense to me.

Following is the conclusion to what we will do to the origin implementation:
```python
def wrapped_handler(interface, handler, cleanup=None):
    ...
    if handler:
        def _handler(widget, *args, **kwargs):    # we need to remove `widget` parameter because it is unused
            ...
            result = handler(interface, *args, **kwargs)
            ...
        ...
```
then we will need to remove the passed argument `'Widget'` into the `self.btn.on_press('Widget', a=1)` to be have `self.btn.on_press(a=1)`

@freakboy3742 please confirm this if accepted. I will close this PR and open new one if this conclusion accepted.

Comment (freakboy3742) [reply src/core/tests/widgets/test_button.py]: I agree that the final usage (`self.btn.on_press(a=1)`) is the desired outcome, but the test *results* shouldn't change (i.e., the interface form of `self.btn` should be the `widget` argument of the handler that is actually invoked). What changes are needed to `wrapped_handler` to make that happen is an exercise for you to confirm - but I *think* you're on the right track.
